securedrop_client/gui/widgets.py

```diff
@@ -1192,6 +1192,56 @@ def __init__(self, source: Source):
         self.setStyleSheet(self.css)
         self.setFixedSize(QSize(20, 20))
 
+    def disable(self):
+        """
+        Disable the widget.
+
+        Disable toggle by setting checkable to False. Unfortunately,
+        disabling toggle doesn't freeze state, rather it always
+        displays the off state when a user tries to toggle. In order
+        to save on state we update the icon's off state image to
+        display on (hack).
+        """
+        self.setCheckable(False)
+        if self.source.is_starred:
+            self.set_icon(on='star_on.svg', off='star_on.svg')
+
+        try:
+            while True:
```

Comment:
won't this disconnect many times? why can't we just disconnect once? it seems like an exception has to be raised in order for this to exit the loop?

Reply:
It will disconnect all the connections of `self.on_toggle_offline` that have been made. There should only ever be one now. And yes, the exception raised when no such connection exists is what exits the loop.